pkg/dynchan/pipechan.go

```diff
@@ -0,0 +1,74 @@
+/*
+ * Minio Client (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dynchan
+
+// Dynamically sized logical channel: a pipe which never blocks even when
+// it receives too many elements. Memory consumption is increased and decresed
+// on the fly following the number of elements.
+func PipeChan(capacity int) (inputCh, outputCh chan interface{}) {
+
+	// A set of channels which store all elements received from input
+	channels := make(chan chan interface{}, 1000)
+
+	inputCh = make(chan interface{}, capacity)
+
+	// A goroutine which receives elements from inputCh and creates
+	// new channels when needed.
+	go func() {
+		// Create the first channel
+		currCh := make(chan interface{}, capacity)
+		channels <- currCh
+
+		for elem := range inputCh {
+			// Prepare next channel with a double capacity when
+			// half of the current channel is already filled.
+			if len(currCh) >= cap(currCh)/2 {
+				close(currCh)
+				currCh = make(chan interface{}, cap(currCh)*2)
+				channels <- currCh
+			}
+			// Prepare next channel with half capacity when
+			// current channel is 1/4 filled
+			if len(currCh) >= capacity && len(currCh) <= cap(currCh)/4 {
+				close(currCh)
+				currCh = make(chan interface{}, cap(currCh)/2)
+				channels <- currCh
+			}
+			// Send element to current channel
+			currCh <- elem
+		}
+
+		close(currCh)
+		close(channels)
+	}()
+
+	// Copy elements from infinite channel set to the output
+	outputCh = make(chan interface{}, capacity)
+	go func() {
+		for {
+			currCh, ok := <-channels
+			if !ok {
+				break
+			}
+			for v := range currCh {
+				outputCh <- v
+			}
+		}
+		close(outputCh)
+	}()
```

Comment:
> This function helps in not loosing any events? does it minimize?

It does minimize. There is no generic way to totally avoid losing events if sender doesn't wait at all for blocked channels.

If the developer chooses 1000 as a starting channel size and the sender was able to send 1500 events before getting scheduled  by the scheduler, 500 events will be lost because we won't get a chance to grow our channel.